src/Features/CSharp/Portable/UseRecursivePatterns/AnalyzedNodeFactory.cs

```diff
@@ -0,0 +1,428 @@
+                IPatternOperation op => VisitPatternOperation(input, op),
+                IIsTypeOperation op => new Type(VisitInput(op.ValueOperand), op.TypeOperand),
+                IIsPatternOperation op => VisitPatternOperation(VisitInput(op.Value), op.Pattern),
+                IPatternCaseClauseOperation op => VisitBinary(op.Pattern, op.Guard, disjunctive: false, input: null),
+                ISwitchExpressionArmOperation op => VisitBinary(op.Pattern, op.Guard, disjunctive: false, input: null),
+                IUnaryOperation { OperatorKind: UnaryOperatorKind.Not } op => VisitNotOperator(op),
+                var op => VisitInput(op)
+            };
+        }
+
+        [return: NotNullIfNotNull("operation")]
+        private static Evaluation? VisitInput(IOperation? operation)
+        {
+            Debug.Assert(!(operation is IConditionalAccessInstanceOperation));
+
+            if (operation is null)
+                return null;
+
+            return VisitNode(operation) ?? new OperationEvaluation(operation);
+
+            static Evaluation? VisitNode(IOperation operation)
+            {
+                return operation switch
+                {
+                    IMemberReferenceOperation op when IsBaseExpressionMemberAccess(op) => null,
+                    ILocalReferenceOperation op => new Variable(input: null, op.Local, op.Syntax),
+                    IFieldReferenceOperation op => new MemberEvaluation(VisitInput(op.Instance), op),
+                    IPropertyReferenceOperation op when op.Property.IsIndexer => VisitIndexerReference(op),
+                    IPropertyReferenceOperation op => VisitNullableTypeMemberAccess(op) ??
+                                                      new MemberEvaluation(VisitInput(op.Instance), op),
+                    IConditionalAccessOperation op => VisitConditionalAccess(VisitInput(op.Operation), op.WhenNotNull),
+                    IConversionOperation op when op.Conversion.IsUserDefined => null,
+                    IConversionOperation op when op.Conversion.IsImplicit => VisitInput(op.Operand),
+                    IConversionOperation op => new Type(VisitInput(op.Operand), op.Type),
+                    _ => null,
+                };
+            }
+
+            static bool IsBaseExpressionMemberAccess(IMemberReferenceOperation operation)
+            {
+                // We cannot use `base` as an standalone expression. So if we encounter `base.x.y.z` we
+                // will walk down until the very last member reference, but not the `base` node itself,
+                // resulting to the following graph:
+                //
+                //  base.x <- y <- z
+                //
+                return operation.Instance is IInstanceReferenceOperation instance &&
+                       instance.ReferenceKind == InstanceReferenceKind.ContainingTypeInstance &&
+                       instance.Syntax.IsKind(SyntaxKind.BaseExpression);
```

Comment:
This is ok as long as we're in csharp layer. any file that doesn't start with CSharp should go to the shared layer then we'll have to make this non-static and use overrides.